### PR TITLE
Feature/min withdrawal amount

### LINKED ISSUE #7302: Add minimum withdrawal amount
https://github.com/vegaprotocol/specs/pull/1508

### PR DESCRIPTION
**Pull request by jeremyletang**
close #7302

### CODE REVIEW

File: core/banking/erc20.go [hunk 1]
```diff
@@ -28,6 +28,7 @@ import (
 
 var (
 	ErrInvalidWithdrawalReferenceNonce       = errors.New("invalid withdrawal reference nonce")
+	ErrWithdrawalAmountUnderMinimalRequired  = errors.New("invalid withdrawal amount under minimal required")
```
Comment (candida-d) [inline core/banking/erc20.go:31]:
```suggestion
	ErrWithdrawalAmountUnderMinimalRequired  = errors.New("invalid withdrawal, amount under minimum required")
```

File: core/banking/erc20.go [hunk 2]
```diff
@@ -197,6 +198,22 @@ func (e *Engine) WithdrawERC20(
...
 	}
 
+	// check for minimal amount reached
+	quantum := asset.Type().Details.Quantum
+	// no reason this would produce an error
+	minAmount, _ := num.UintFromDecimal(quantum.Mul(e.minWithdrawQuantumMultiple))
+
+	// now verify amount
+	if amount.LT(minAmount) {
+		e.log.Debug("cannot withdraw funds, less than minimal amount requested to withdraw",
```
Comment (candida-d) [inline core/banking/erc20.go:208]:
```suggestion
		e.log.Debug("cannot withdraw funds, the request is less than minimum withdrawal amount",
```

File: core/netparams/keys.go [hunk 3]
```diff
@@ -287,4 +288,5 @@ var AllKeys = map[string]struct{}{
 	SpamProtectionMaxBatchSize:                               {},
 	MarkPriceUpdateMaximumFrequency:                          {},
 	ValidatorPerformanceScalingFactor:                        {},
+	SpamProtectionMinimiumWithdrawalQuantumMultiple:          {},
```
Comment (candida-d) [inline core/netparams/keys.go:291]:
```suggestion
	SpamProtectionMinimumWithdrawalQuantumMultiple:          {},
```

Comment (candida-d) [reply core/netparams/keys.go:291]: fixes spelling of minimum

File: core/netparams/keys.go [hunk 4]
```diff
@@ -110,13 +110,14 @@ const (
...
-	SpamProtectionMinTokensForDelegation = "spam.protection.delegation.min.tokens"
-	SpamProtectionMaxBatchSize           = "spam.protection.max.batchSize"
+	SpamProtectionMaxVotes                          = "spam.protection.max.votes"
+	SpamProtectionMinTokensForVoting                = "spam.protection.voting.min.tokens"
+	SpamProtectionMaxProposals                      = "spam.protection.max.proposals"
+	SpamProtectionMinTokensForProposal              = "spam.protection.proposal.min.tokens"
+	SpamProtectionMaxDelegations                    = "spam.protection.max.delegations"
+	SpamProtectionMinTokensForDelegation            = "spam.protection.delegation.min.tokens"
+	SpamProtectionMaxBatchSize                      = "spam.protection.max.batchSize"
+	SpamProtectionMinimiumWithdrawalQuantumMultiple = "spam.protection.minimumWithdrawalQuantumMultiple"
```
Comment (candida-d) [inline core/netparams/keys.go:120]:
```suggestion
	SpamProtectionMinimumWithdrawalQuantumMultiple = "spam.protection.minimumWithdrawalQuantumMultiple"
```

File: core/protocol/all_services.go [hunk 5]
```diff
@@ -641,6 +641,10 @@ func (svcs *allServices) setupNetParameters(powWatchers []netparams.WatchParam)
 			Param:   netparams.TransferMinTransferQuantumMultiple,
 			Watcher: svcs.banking.OnMinTransferQuantumMultiple,
 		},
+		{
+			Param:   netparams.SpamProtectionMinimiumWithdrawalQuantumMultiple,
```
Comment (candida-d) [inline core/protocol/all_services.go:645]:
```suggestion
			Param:   netparams.SpamProtectionMinimumWithdrawalQuantumMultiple,
```

**Review by candida-d — COMMENTED** [review]
spelling fixes and suggestions for changes to error messages

**Review by ValentinTrinque — APPROVED** [review]

File: core/netparams/defaults.go [hunk 6]
```diff
@@ -162,13 +162,13 @@ func defaultNetParams() map[string]value {
...
-		SpamProtectionMaxDelegations:         NewInt(gteI1).Mutable(true).MustUpdate("390"),
-		SpamProtectionMinTokensForDelegation: NewDecimal(gteD1).Mutable(true).MustUpdate("1000000000000000000"),
-
+		SpamProtectionMaxVotes:                         NewInt(gteI1).Mutable(true).MustUpdate("3"),
+		SpamProtectionMinTokensForVoting:               NewDecimal(gteD1).Mutable(true).MustUpdate("100000000000000000000"),
+		SpamProtectionMaxProposals:                     NewInt(gteI1).Mutable(true).MustUpdate("3"),
+		SpamProtectionMinTokensForProposal:             NewDecimal(gteD1).Mutable(true).MustUpdate("100000000000000000000000"),
+		SpamProtectionMaxDelegations:                   NewInt(gteI1).Mutable(true).MustUpdate("390"),
+		SpamProtectionMinTokensForDelegation:           NewDecimal(gteD1).Mutable(true).MustUpdate("1000000000000000000"),
+		SpamProtectionMinimumWithdrawalQuantumMultiple: NewDecimal(DecimalGT(num.MustDecimalFromString("0")), DecimalLT(num.MustDecimalFromString("10e6"))).Mutable(true).MustUpdate("10"),
```
Comment (ze97286) [inline core/netparams/defaults.go:171]: SpamProtectionMinimumWithdrawalQuantumMultiple: NewDecimal(gt0, DecimalLT(num.MustDecimalFromString("10e6"))).Mutable(true).MustUpdate("10"),

also, is this meant to be 1e7? or did you mean 1e6?

Comment (jeremyletang) [reply core/netparams/defaults.go:171]: This meant to be "10^6" per what Klaus asked me?

Comment (ze97286) [reply core/netparams/defaults.go:171]: 10^6 is 1e6 not 10e6

**Pull request merged**